openupgrade_scripts/scripts/hr_holidays/15.0.1.5/pre-migration.py

```diff
@@ -0,0 +1,240 @@
+from openupgradelib import openupgrade
+
+
+def _fast_fill_hr_leave_employee_company_id(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE hr_leave
+        ADD COLUMN IF NOT EXISTS employee_company_id integer""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE hr_leave hl
+        SET employee_company_id = empl.company_id
+        FROM hr_employee empl
+        WHERE empl.company_id IS NOT NULL AND hl.employee_id = empl.id""",
+    )
+
+
+def _map_hr_leave_state(env):
+    openupgrade.map_values(
+        env.cr,
+        "state",
+        "state",
+        [("cancel", "refuse")],
+        table="hr_leave",
+    )
+
+
+def _map_hr_leave_allocation_approver_id(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE hr_leave_allocation
+        ADD COLUMN IF NOT EXISTS approver_id integer""",
+    )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE hr_leave_allocation
+        SET approver_id = CASE
+            WHEN second_approver_id IS NOT NULL THEN second_approver_id
+            ELSE first_approver_id END
+        WHERE state in ('refuse', 'validate')""",
+    )
+
+
+def _map_hr_leave_allocation_state(env):
+    openupgrade.logged_query(
+        env.cr,
+        """UPDATE hr_leave_allocation
+        SET state = 'confirm'
+        WHERE state = 'validate1'""",
+    )
+
+
+def _convert_datetime_to_date_hr_leave_allocation_date_from(env):
+    openupgrade.logged_query(
```

Comment:
You have to first rename the existing columns, then create a new ones, and then transfer data converted to date.